### PR TITLE
perf(dspark): add load-aware compact verification capacity

### PR DESCRIPTION
**Pull request by voipmonitor**
## Summary

Add opt-in load-aware physical draft-depth control for DSpark/DFlash while keeping the configured logical width and target acceptance semantics unchanged.

At low load the controller retains exact fixed-K5 behavior. At the configured activation knee it selects a smaller physical depth, uses compact variable-length verification, and keeps the path on full CUDA graphs. TP rank 0 computes and broadcasts one canonical capacity vector so all ranks replay the same padded shape without a host-wide synchronization.

## Dependency

Merge #108 and #105 first. #108 removes a duplicate sparse-indexer helper that otherwise makes the combined changed-file lint set fail; #105 provides the corrected DFlash prefix-cache, request-slot, graph-lifetime, and padding behavior consumed by this capacity path. The PR still targets `dev/fathomless-firmament`; after both prerequisites merge, GitHub will reduce this diff to the capacity feature.

## Behavior and scope

- disabled by default
- fixed K remains unchanged below the load knee
- compact variable-length verification activates only at the configured load threshold
- one TP-canonical capacity vector is broadcast before asynchronous D2H transfer
- dynamic-depth warmup and full-graph captures cover every enabled physical depth
- explicit validation rejects inconsistent capacity configuration
- profiling utility and focused unit coverage are included

## Measured TP2 Lucifer results

Fixed K5 baseline versus load-aware capacity:

| Concurrency | Fixed K5 tok/s | Load-aware tok/s |
|---:|---:|---:|
| 1 | 237.2 | 239.5 |
| 8 | 770.9 | 764.5 |
| 16 | 1186.1 | 1180.1 |
| 32 | 1764.6 | 1799.6 |
| 64 | 2614.6 | 2718.2 |

Three-run C64 mean was 2725.1 tok/s with SD 11.2. Coding median was 323.0 tok/s, mean 327.5, maximum 359.2. A clean-cache C1/C64/recovery sequence measured 245.2 / 2700.5 / 241.3 tok/s; a 120-second C64 run measured 2751.4 tok/s.

## Rejected approaches

- Masking unused slots while retaining fixed physical width does not save CUTLASS target compute.
- Always-on compact verification regresses C1.
- Global synchronization and event-handle reuse were rejected because they either serialize the hot path or are unsafe across asynchronous replays.
- B12X DMA and NCCL tuning are independent transport experiments and are not part of this change.

## Validation

- 68 focused DSpark correctness and capacity tests passed
- 15 B12X sparse-indexer tests passed with current DCP and KV-cache fixtures
- clean canonical integration stack: 233 passed, 714 skipped
- Ruff lint and format checks pass for the complete changed Python set
- `git diff --check` passes

This is extracted from the old mixed #88 branch so the capacity optimization can be reviewed and merged independently.

## Summary by CodeRabbit

* **New Features**
  * Added DSpark capacity-based speculative decoding with configurable varlen or masked verification modes.
  * Added confidence-based draft-token budgeting, online calibration, dynamic draft depth, and SPS curve support.
  * Added automatic SPS profiling and a standalone tool for generating SPS configuration curves.
  * Added new command-line and environment settings for DSpark capacity behavior.
  * Added DeepSeek and Qwen DSpark confidence-head support.
  * Added a GSM8K evaluation configuration for DeepSeek V4 DSpark.

* **Bug Fixes**
  * Improved CUDA graph dispatch, prefix-cache handling, padding masks, sampling bounds, and deterministic fallback seeds.
  * Improved sparse attention handling for variable-length and non-causal workloads.

* **Tests**
  * Expanded coverage for DSpark configuration, capacity scheduling, CUDA graphs, attention metadata, sampling, and prefix caching.

### CODE REVIEW

File: benchmarks/profile_dspark_sps_curve.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+def curve_breakpoints(
+    ms_per_step: dict[int, float], overhead_ms: float
+) -> list[list[float]]:
+    """Convert per-capture-size step times into ``dspark_sps_curve``
+    breakpoints, one per capture size. The scheduler's table linearly
+    interpolates between them (and clamps at the ends)."""
+    return [
+        [size, round(1000.0 / (ms_per_step[size] + overhead_ms), 3)]
+        for size in sorted(ms_per_step)
+    ]
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply benchmarks/profile_dspark_sps_curve.py:97]: Fixed in edc49763b: the CLI now rejects non-positive iterations and negative warmup or overhead before profiling.

File: tests/v1/worker/test_mixed_warmup_gate.py [hunk 2]
```diff
@@ -6,13 +6,35 @@
...
+        (6, 87.79),
+        (12, 71.18),
+        (24, 67.71),
+        (48, 58.52),
+        (96, 51.29),
+        (192, 40.33),
+        (384, 23.72),
+    ]
+
+    assert _derive_dspark_draft_token_budget(curve, max_draft_depth=5) == 160
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tests/v1/worker/test_mixed_warmup_gate.py:35]: Fixed in edc49763b: the test now has a larger lower-half SPS drop while retaining the expected upper-half knee and budget.

File: vllm/model_executor/models/qwen3_dspark.py [hunk 3]
```diff
@@ -67,8 +69,33 @@ def bias(self, markov_embed: torch.Tensor, logits_processor) -> torch.Tensor:
...
+            1,
+            bias=bias,
+            return_bias=False,
+            params_dtype=torch.float32,
+            prefix=maybe_prefix(prefix, "proj"),
+        )
+
+    def forward(self, hidden: torch.Tensor, markov_embed: torch.Tensor) -> torch.Tensor:
+        x = torch.cat([hidden, markov_embed], dim=-1).float()
+        return self.proj(x).squeeze(-1)
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/model_executor/models/qwen3_dspark.py:103]: Fixed in edc49763b: the confidence head now honors include_markov, and a regression test covers both projection shapes.

File: vllm/v1/attention/backends/mla/indexer.py [hunk 4]
```diff
@@ -269,6 +272,50 @@ def get_max_prefill_buffer_size(vllm_config: VllmConfig):
...
+    return bool(
+        spec_config is not None
+        and spec_config.use_dspark()
+        and spec_config.dspark_capacity_verification_mode == "varlen"
+        and (
+            spec_config.dspark_confidence_threshold > 0.0
+            or spec_config.dspark_budget_frac < 1.0
+            or spec_config.dspark_sps_curve is not None
+        )
+    )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/attention/backends/mla/indexer.py:302]: Fixed in edc49763b: CUDA-graph support and runtime varlen enablement now both require hardware support and explicit DSpark capacity opt-in; the gate is covered by a regression test.

File: vllm/v1/worker/gpu/model_runner.py [hunk 5]
```diff
@@ -1098,7 +1144,41 @@ def prepare_inputs(
...
+        num_tokens = input_batch.num_tokens
+        num_tokens_after_padding = input_batch.num_tokens_after_padding
+        assert 0 < num_tokens <= num_tokens_after_padding, (
+            f"Batch has {num_tokens} tokens after trimming but was dispatched "
+            f"for {num_tokens_after_padding}"
         )
+        if self.verification_capacity_manager is None and (
+            envs.VLLM_MOE_SKIP_PADDING or input_batch.num_draft_tokens > 0
+        ):
+            self.input_buffers.is_padding[:num_tokens].fill_(False)
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/worker/gpu/model_runner.py]: Fixed in edc49763b. The active slice is now reset unconditionally before the capacity manager runs. That clears stale plain-batch state while preserving padding deliberately applied later by masked capacity; resetting after the manager would erase its skip mask.

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 6]
```diff
@@ -146,9 +269,118 @@ def _sample_sequential(self, num_reqs: int, head_hidden: torch.Tensor) -> None:
...
+                num_reqs,
+                n_spec,
+                self._runtime_num_reqs_for_capacity,
+                self.draft_token_survival_probs,
+                self.capacity_budget_frac,
+                sps_table=self.sps_table,
+                confidence_temperature=capacity_temperature,
+            )
+        else:
+            self.draft_token_capacity[:num_reqs].fill_(n_spec)
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/worker/gpu/spec_decode/dspark/speculator.py:306]: Fixed in edc49763b. Online STS tracks proposal validity per request-state slot, invalidates bypass transitions, rejects profile/dummy/unaligned proposals, and consumes each staged proposal once. Regression coverage verifies invalid and repeated records leave calibration unchanged.

**Comment by voipmonitor** [comment]
Canonical-stack update: PR #108 is now a git ancestor of this branch, so the sparse-indexer dedup/lint prerequisite is explicit. The full #107 changed-file set now passes `ruff check`, `ruff format --check`, and `git diff --check` against current FF. Merge #108 before #107; #105 remains the DSpark correctness prerequisite.

**Comment by voipmonitor** [comment]
## Canonical FF + B12X merge handoff (2026-07-17)

Pinned bases, rechecked immediately before this handoff:

- vLLM `dev/fathomless-firmament`: `e4879e195dd9ecf23661803f6775c6a847a18a72`
- B12X `master`: `115926eb6452d1e8b849225d7af112e2bb184030`

### 1. Merge B12X first

1. [lukealonso/b12x#28](https://github.com/lukealonso/b12x/pull/28)
2. [lukealonso/b12x#32](https://github.com/lukealonso/b12x/pull/32)
3. [lukealonso/b12x#34](https://github.com/lukealonso/b12x/pull/34)
4. [lukealonso/b12x#35](https://github.com/lukealonso/b12x/pull/35)

Then close superseded [lukealonso/b12x#33](https://github.com/lukealonso/b12x/pull/33). It is still draft/conflicting because this account cannot close another author’s PR; #35 is its clean replacement.

This exact B12X order was synthesized cleanly and validated with **122 passed, 1 skipped**. All four PRs are non-draft, mergeable, clean, and have no unresolved review threads.

### 2. Merge vLLM into FF in this exact order

1. #108
2. #97
3. #103
4. #98
5. #99
6. #100
7. #101
8. #102
9. #104
10. #105
11. #107
12. #106

The exact sequence was rebuilt from the pinned FF head with no manual conflict resolution. Final validation: **248 passed, 714 skipped**, Ruff lint and format clean across all **86 changed Python files**, and `git diff --check` clean. There are no unresolved review threads on #97-#108, every PR is non-draft and GitHub reports every PR mergeable. The remaining `UNSTABLE` labels on a few independent PRs are queued repository `pre-commit` jobs, not failed checks.

### Merge strategy

Use normal merge commits for the stacked pairs:

- #97 before #103
- #108 and #105 before #107

If FF policy requires squash or rebase merges, stop after rewriting #97, #108, or #105 and rebase the dependent PR (#103 or #107) onto the new FF head before continuing. Otherwise GitHub cannot recognize the tested ancestry and may show duplicate changes.

The old mixed #88 and superseded #92/#96 are closed; their canonical replacements are the PRs above.

**Comment by voipmonitor** [comment]
Superseded by #109. The final code was rebuilt directly on current dev/fathomless-firmament as three clean commits: correctness, optional FP8 draft head, and optional load-aware capacity. #109 preserves the validated behavior while removing the stacked merge history and excluding the unrelated #108 indexer cleanup.
